### PR TITLE
Report deregister event if read snapshot file

### PR DESCRIPTION
**Pull request by Daydreamer-ia**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

XXXXX
#10406
## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/manager/impl/PersistentIpPortClientManager.java [hunk 1]
```diff
@@ -131,6 +131,7 @@ public Map<String, IpPortBasedClient> showClients() {
     public void loadFromSnapshot(ConcurrentMap<String, IpPortBasedClient> clients) {
         ConcurrentMap<String, IpPortBasedClient> oldClients = this.clients;
         this.clients = clients;
+        oldClients.forEach((clientId, client) -> client.release());
```
Comment (Lin-1997) [inline naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/manager/impl/PersistentIpPortClientManager.java:134]: 似乎有问题，调用链是：
PersistentInstanceSnapshotOperation#loadSnapshot----snapshotClient.init();
-->
clientManager.loadFromSnapshot(snapshot);
-->
oldClients.forEach((clientId, client) -> client.release());

上面的snapshotClient.init();如果是clientBeat的形式，会调：
HealthCheckReactor#scheduleCheck----futureMap.computeIfAbsent(task.taskKey(), xxx)
这task还没有释放，所以不会再加进去

但是后面oldClients.forEach((clientId, client) -> client.release());会释放掉，导致没有了clientBeat的task

Comment (Daydreamer-ia) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/manager/impl/PersistentIpPortClientManager.java]: i get it

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java [hunk 2]
```diff
@@ -343,18 +347,111 @@ protected InputStream dumpSnapshot() {
...
+            for (Service service : aliveInstanceServices) {
+                Service singleton = ServiceManager.getInstance().getSingleton(service);
+                InstancePublishInfo newInstanceInfo = newInstanceInfoMap.get(singleton);
+                if (oldPublishedService.contains(singleton)) {
+                    // update if necessary
+                    InstancePublishInfo oldInstanceInfo = client.getInstancePublishInfo(singleton);
+                    if (oldInstanceInfo != null && !newInstanceInfo.equals(oldInstanceInfo)) {
+                        client.addServiceInstance(singleton, newInstanceInfo);
+                        NotifyCenter.publishEvent(new ClientOperationEvent.ClientRegisterServiceEvent(singleton, client.getClientId()));
+                        Loggers.RAFT.info("[SNAPSHOT-DATA-ADD] service={}, instance={}", service, newInstanceInfo);
```
Comment (JianweiWang) [inline naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java:399]: 日志这里最好能区分下ADD 和 UPDATE

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java [hunk 3]
```diff
@@ -343,18 +347,111 @@ protected InputStream dumpSnapshot() {
...
+                    // add
+                    client.addServiceInstance(singleton, newInstanceInfo);
+                    NotifyCenter.publishEvent(new ClientOperationEvent.ClientRegisterServiceEvent(singleton, client.getClientId()));
+                    Loggers.RAFT.info("[SNAPSHOT-DATA-ADD] service={}, instance={}", service, newInstanceInfo);
+                }
+            }
+            // remove dead instance
+            for (Service service : oldPublishedService) {
+                if (!aliveInstanceServices.contains(service)) {
+                    Loggers.RAFT.info("[SNAPSHOT-DATA-REMOVE] service={}, instance={}", service, client.getInstancePublishInfo(service));
```
Comment (JianweiWang) [inline naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java:411]: 日志放到操作完成之后，和ADD、UPDATE保持一致

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java (hunk 2 above)
Comment (Daydreamer-ia) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java:399]: i get it

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java (hunk 3 above)
Comment (Daydreamer-ia) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java:411]: i get it

**Review by pixystone — CHANGES_REQUESTED** [review]
Be aware of the performance.

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/manager/impl/PersistentIpPortClientManager.java [hunk 4]
```diff
@@ -133,4 +133,25 @@ public void loadFromSnapshot(ConcurrentMap<String, IpPortBasedClient> clients) {
         this.clients = clients;
         oldClients.clear();
     }
+
+    /**
+     * add client directlt.
```
[the remark this answers is not in the document]
Comment (Daydreamer-ia) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/manager/impl/PersistentIpPortClientManager.java:138]: i get it

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java [hunk 5]
```diff
@@ -343,18 +347,111 @@ protected InputStream dumpSnapshot() {
...
+            Set<Service> aliveInstanceServices = newInstanceInfoMap.keySet();
+            // add or update existed service
+            for (Service service : aliveInstanceServices) {
+                Service singleton = ServiceManager.getInstance().getSingleton(service);
+                InstancePublishInfo newInstanceInfo = newInstanceInfoMap.get(singleton);
+                if (oldPublishedService.contains(singleton)) {
+                    // update if necessary
+                    InstancePublishInfo oldInstanceInfo = client.getInstancePublishInfo(singleton);
+                    if (oldInstanceInfo != null && !newInstanceInfo.equals(oldInstanceInfo)) {
+                        client.addServiceInstance(singleton, newInstanceInfo);
```
[the remark this answers is not in the document]
Comment (Daydreamer-ia) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java:397]: i get it

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Daydreamer-ia) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java]: i get it

File: naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java [hunk 6]
```diff
@@ -343,18 +347,111 @@ protected InputStream dumpSnapshot() {
...
-            ConcurrentHashMap<String, IpPortBasedClient> snapshot = new ConcurrentHashMap<>(newData.size());
+            Collection<String> oldClientIds = clientManager.allClientId();
+            // add or update
             for (Map.Entry<String, ClientSyncData> entry : newData.entrySet()) {
-                IpPortBasedClient snapshotClient = new IpPortBasedClient(entry.getKey(), false);
-                snapshotClient.setAttributes(entry.getValue().getAttributes());
-                snapshotClient.init();
-                loadSyncDataToClient(entry, snapshotClient);
-                snapshot.put(entry.getKey(), snapshotClient);
+                if (oldClientIds.contains(entry.getKey())) {
```
[the remark this answers is not in the document]
Comment (Daydreamer-ia) [reply naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java:353]: i get it

**Comment by Lin-1997** [comment]
can we just throw away the old data and load the new data? I think this is more suitable for the semantics of installing snapshot.
i.e., disconnect all old clients, and then publish all clients from snapshot, instead of updating/deleting existing data

**Comment by Daydreamer-ia** [comment]
> can we just throw away the old data and load the new data? I think this is more suitable for the semantics of installing snapshot. i.e., disconnect all old clients, and then publish all clients from snapshot, instead of updating/deleting existing data

@KomachiSion He thinks the old way to load snapshot is better. How can i do to solve it?

**Comment by Lin-1997** [comment]
> > can we just throw away the old data and load the new data? I think this is more suitable for the semantics of installing snapshot. i.e., disconnect all old clients, and then publish all clients from snapshot, instead of updating/deleting existing data
> 
> @KomachiSion He thinks the old way to load snapshot is better. How can i do to solve it?

My idea is not suitable in some cases. If the follower is just disconnected to the leader right before the leader saves snapshot, and recovers soon. Throw away all the old data also seems to be unreasonable.

**Comment by Lin-1997** [comment]
how about this:

we construct snapshot data in `PersistentClientOperationServiceImpl` and handle it in `PersistentIpPortClientManager`.

first, `PersistentClientOperationServiceImpl` only constructs `ConcurrentHashMap<String, IpPortBasedClient> snapshot` and doesn't modify data in other `manager` or `index`, i.e., remove L368-369.

https://github.com/alibaba/nacos/blob/51903edd82a3f244a5e21f173f011485298f348f/naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java#L357-L371

Also we doesn't need to init snapshotClient in L350, because we are going to regenerate the client in `PersistentIpPortClientManager`.

https://github.com/alibaba/nacos/blob/51903edd82a3f244a5e21f173f011485298f348f/naming/src/main/java/com/alibaba/nacos/naming/core/v2/service/impl/PersistentClientOperationServiceImpl.java#L344-L355

then, we handle the snapshot in `PersistentIpPortClientManager` in the follow function:

https://github.com/alibaba/nacos/blob/51903edd82a3f244a5e21f173f011485298f348f/naming/src/main/java/com/alibaba/nacos/naming/core/v2/client/manager/impl/PersistentIpPortClientManager.java#L131-L135

like this:

```
public void loadSnapshot(ConcurrentMap<String, IpPortBasedClient> snapshot) {
    // disconnect expired clients
    Set<String> shouldDelete = new HashSet<>(clients.keySet());
    shouldDelete.removeAll(snapshot.keySet());
    shouldDelete.forEach(this::clientDisconnected); // TODO should be enough, needed to be confirmed

    // load snapshot clients
    snapshot.forEach((clientId, snapshotClient) -> {
        Client existingClient;
        if (contains(clientId)) {
            existingClient = getClient(clientId);
            removeExpiredInstances(existingClient, snapshotClient);
        } else {
            clientConnected(clientId, snapshotClient.getClientAttributes()); // TODO deep copy of clientAttributes
            existingClient = getClient(clientId);
            ((IpPortBasedClient) existingClient).init();
        }

        loadSnapshotInstances(existingClient, snapshotClient);
    });
    snapshot.clear();
}

private void removeExpiredInstances(Client existingClient, Client snapshotClient) {
    Set<Service> shouldDelete = new HashSet<>(existingClient.getAllPublishedService());
    shouldDelete.removeAll(snapshotClient.getAllPublishedService());
    shouldDelete.forEach(service -> {
        Service singleton = ServiceManager.getInstance().getSingleton(service);
        existingClient.removeServiceInstance(singleton);
        NotifyCenter.publishEvent(new ClientOperationEvent
                .ClientDeregisterServiceEvent(singleton, existingClient.getClientId()));
    });
}

private void loadSnapshotInstances(Client existingClient, Client snapshotClient) {
    snapshotClient.getAllPublishedService().forEach(service -> {
        Service singleton = ServiceManager.getInstance().getSingleton(service);
        existingClient.addServiceInstance(singleton, snapshotClient.getInstancePublishInfo(singleton));
        NotifyCenter.publishEvent(new ClientOperationEvent
                .ClientRegisterServiceEvent(singleton, existingClient.getClientId()));
    });
}
```

**Comment by pixystone** [comment]
There is only a single thread applying the raft log of every group in sequence. So I think the in-place modifying method is suitable here in order that we don't need to consider the thread-safe problem.

**Comment by Daydreamer-ia** [comment]
> There is only a single thread applying the raft log of every group in sequence. So I think the in-place modifying method is suitable here in order that we don't need to consider the thread-safe problem.

Yes, I have done in place instead of replacing them in full

**Comment by Daydreamer-ia** [comment]
> https://github.com/alibaba/nacos/blob/8cd5a5ea0bfb7014523eb19ca72e28282a9155a2/naming/src/main/java/com/alibaba/nacos/naming/core/v2/metadata/NamingMetadataManager.java#L193-L211
> 
> may lose the true Service singleton, since it is load from snapshot, while the singleton in menory doesn't change. Also have not clean the expried ones.
> 
> may also has to call ServiceStorage#removeData of those expried services. But we can not call this function in MetadataManager, my idea is make a new type of event, and publish in ServiceManager#removeSingleton.

i see. You mean that there are some outdated service index in ServiceStorage#serviceDataIndexes without clean? The empty service will be clean by EmptyServiceAutoCleanerV2 actually, so there is unnecessary to make a new type of event to clean.

**Comment by pixystone** [comment]
LGTM

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**